Add option to cache .vsix and .theia artifacts during build

amisevsk opened the pull request:
### What does this PR do?
Add `cache_artifacts.sh` and some optional build stages in the Dockerfile / Dockerfile.rhel which can be used to build a plugin registry that includes .vsix and .theia files, to support running Che in an air-gapped scenario.

PR is ready to merge.

nickboldt commented:
Are you also caching SVG files (icons) from eclipse.org?

amisevsk commented:
Should add caching of icon files as well before merging PR.

amisevsk commented:
@l0rd good point, I'll rework the PR.

nickboldt commented:
> You can use an offline-mode build ARG. Or you can add 2 new stages in the multi stage build:

ARGs don't work in Brew. Would prefer multistage, which I can do.

amisevsk commented:
Update: I reworked the `Dockerfile.caching` into a couple additional stages on the regular docker build. Now to build the regular registry, you would use
```
docker build <args> --target registry
```
and to build the offline registry, you would use
```
docker build <args> --target offline-registry
```

WDYT?

amisevsk commented:
The main downside with this approach is that the offline registry is what you get if you just `docker build .`, but it's also kind of unavoidable, since moving the offline stage earlier means it's built every time and then discarded.

amisevsk commented:
> Should add caching of icon files as well before merging PR.

I still don't know a good way of doing this (could download them as well but there's no clear way to update `meta.yaml`s to point to the downloaded images).

nickboldt commented:
> The main downside with this approach is that the offline registry is what you get if you just `docker build .`, but it's also kind of unavoidable, since moving the offline stage earlier means it's built every time and then discarded.

From the perspective of support/product/airgap customers, this is an **upside**. :+1:

nickboldt commented:
> I still don't know a good way of doing this (could download them as well but there's no clear way to update meta.yamls to point to the downloaded images).

I'm OK if image caching support is handled in #221 instead. I think the simplest approach is:
* pull the vsix, look for .png or .svg file, if found use that
* if no vsix or no image found in the vsix archive, pull the referenced image in the meta.yaml
* it none found or specified, fall back to generic/default image (which should also be in the plugin registry if not present in Theia itself)

nickboldt commented:
Looks good. Build log is thankfully verbose (helps with prod build auditing):

```
2019-09-26 19:45:13 URL:https://github-production-release-asset-2e65be.s3.amazonaws.com/134612364/bb0b5900-9113-11e9-878b-e84fe8e3994d?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190926%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190926T194508Z&X-Amz-Expires=300&X-Amz-Signature=4cfd01004fc8f733909c809f9d436b82ab08d00b05585ab1f22a19e21f4bb01f&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dvscode-xml-0.7.0-3205.vsix&response-content-type=application%2Foctet-stream [6163460/6163460] -> "v3/extensions_temp/vscode-xml-0.7.0-3205.vsix" [1]

2019-09-26 19:45:28 URL:https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-vscode/vsextensions/node-debug2/1.31.6/vspackage [630628/630628] -> "v3/extensions_temp/ms-vscode.node-debug2-1.31.6.vsix" [1]
```

Curiously, the created image is HUGE though but maybe that's because you're caching the same artifacts more than once?

```
docker images | grep plugin-registry

che-plugin-registry                                         amisevsk             e1fecee43fce        49 seconds ago       1.35GB
quay.io/eclipse/che-plugin-registry                         nightly              ec996eee40a2        20 hours ago        366MB
quay.io/nickboldt/airgap-che-plugin-registry                20190922-1740        861e537e7bf2        3 days ago           832MB
```

nickboldt commented:
My approach to caching (including images):

```
[jboss@78aab22f877c resources]$ du -sch $(find . -type f -name "*vsix" -o -name "*theia" | sort)
2.3M	./download.jboss.org/jbosstools/adapters/stable/vscode-project-initializer/project_initializer-0.0.9-92.vsix
38M	./download.jboss.org/jbosstools/static/jdt.ls/stable/java-0.38.0-1373.vsix
38M	./download.jboss.org/jbosstools/static/jdt.ls/stable/java-0.43.0-1473.vsix
38M	./download.jboss.org/jbosstools/static/jdt.ls/stable/java-0.45.0-1523.vsix
38M	./download.jboss.org/jbosstools/static/jdt.ls/stable/java-0.46.0-1549.vsix
19M	./download.jboss.org/jbosstools/vscode/stable/vscode-apache-camel/vscode-apache-camel-0.0.16-37.vsix
24M	./download.jboss.org/jbosstools/vscode/stable/vscode-wsdl2rest/vscode-wsdl2rest-0.0.9-16.vsix
17M	./github.com/Azure/vscode-kubernetes-tools/releases/download/0.1.17/vscode-kubernetes-tools-0.1.17.vsix
13M	./github.com/Azure/vscode-kubernetes-tools/releases/download/1.0.0/vscode-kubernetes-tools-1.0.0.vsix
2.1M	./github.com/Microsoft/vscode-go/releases/download/0.9.2/Go-0.9.2.vsix
9.2M	./github.com/Microsoft/vscode-python/releases/download/2019.3.6558/ms-python-release.vsix
14M	./github.com/Microsoft/vscode-python/releases/download/2019.5.18875/ms-python-release.vsix
35M	./github.com/SonarSource/sonarlint-vscode/releases/download/1.6.0/sonarlint-vscode-1.6.0.vsix
19M	./github.com/camel-tooling/camel-lsp-client-vscode/releases/download/0.0.14/camel-tooling.vscode-apache-camel-0.0.14.vsix
1.9M	./github.com/che-incubator/che-cpptools/releases/download/v1.0/cdt-gdb-vscode-0.0.90.vsix
1.6M	./github.com/che-incubator/che-cpptools/releases/download/v1.0/cdt-vscode-0.0.7.vsix
472K	./github.com/che-incubator/che-cpptools/releases/download/v1.0/llvm-vs-code-extensions.vscode-clangd-0.0.16.vsix
12M	./github.com/che-incubator/ms-code.typescript/releases/download/v1.30.2/che-typescript-language.vsix
12M	./github.com/che-incubator/ms-code.typescript/releases/download/v1.35.1/che-typescript-language-1.35.1.vsix
3.6M	./github.com/che-incubator/vscode-intelephense/releases/download/v1.0.13/bmewburn.vscode-intelephense-client-1.0.13.vsix
2.2M	./github.com/fabric8-analytics/fabric8-analytics-vscode-extension/releases/download/0.0.12/redhat.fabric8-analytics-0.0.12.vsix
1.1M	./github.com/felixfbecker/vscode-php-debug/releases/download/v1.13.0/php-debug.vsix
2.1M	./github.com/microsoft/vscode-go/releases/download/0.11.0/Go-0.11.0.vsix
3.7M	./github.com/microsoft/vscode-java-debug/releases/download/0.16.0/vscode-java-debug-0.16.0.vsix
2.5M	./github.com/microsoft/vscode-java-debug/releases/download/0.19.0/vscode-java-debug-0.19.0.vsix
472K	./github.com/microsoft/vscode-pull-request-github/releases/download/v0.8.0/vscode-pull-request-github-0.8.0.vsix
3.3M	./github.com/redhat-developer/netcoredbg-theia-plugin/releases/download/v0.0.1/netcoredbg_theia_plugin.theia
31M	./github.com/redhat-developer/omnisharp-theia-plugin/releases/download/v0.0.1/omnisharp_theia_plugin.theia
43M	./github.com/redhat-developer/omnisharp-theia-plugin/releases/download/v0.0.2/omnisharp_theia_plugin.theia
45M	./github.com/redhat-developer/omnisharp-theia-plugin/releases/download/v0.0.4/omnisharp_theia_plugin.theia
4.2M	./github.com/redhat-developer/vscode-openshift-tools/releases/download/v0.0.17/openshift-connector-0.0.17-127.vsix
4.5M	./github.com/redhat-developer/vscode-openshift-tools/releases/download/v0.0.19/openshift-connector-0.0.19-183.vsix
4.6M	./github.com/redhat-developer/vscode-openshift-tools/releases/download/v0.0.21/openshift-connector-0.0.21-228.vsix
5.6M	./github.com/redhat-developer/vscode-xml/releases/download/0.3.0/redhat.vscode-xml-0.3.0.vsix
5.6M	./github.com/redhat-developer/vscode-xml/releases/download/0.5.1/redhat.vscode-xml-0.5.1.vsix
5.9M	./github.com/redhat-developer/vscode-xml/releases/download/0.7.0/vscode-xml-0.7.0-3205.vsix
5.8M	./github.com/redhat-developer/vscode-yaml/releases/download/0.3.0/redhat.vscode-yaml-0.3.0.vsix
5.8M	./github.com/redhat-developer/vscode-yaml/releases/download/0.4.0/redhat.vscode-yaml-0.4.0.vsix
16K	./raw.githubusercontent.com/theia-ide/theia
504M	total
```
Your approach:

```
bash-4.2# du -sch $(find . -type f -name "*vsix" -o -name "*theia" | sort)
2.1M	./Go-0.11.0.vsix
2.1M	./Go-0.9.2.vsix
3.6M	./bmewburn.vscode-intelephense-client-1.0.13.vsix
19M	./camel-tooling.vscode-apache-camel-0.0.14.vsix
1.9M	./cdt-gdb-vscode-0.0.90.vsix
1.6M	./cdt-vscode-0.0.7.vsix
12M	./che-typescript-language-1.35.1.vsix
12M	./che-typescript-language.vsix
38M	./java-0.38.0-1373.vsix
38M	./java-0.43.0-1473.vsix
38M	./java-0.45.0-1523.vsix
38M	./java-0.46.0-1549.vsix
38M	./joaompinto.asciidoctor-vscode-2.7.6.vsix
472K	./llvm-vs-code-extensions.vscode-clangd-0.0.16.vsix
14M	./ms-python-release.vsix
9.1M	./ms-python.python-2019.2.5433.vsix
184K	./ms-vscode.node-debug-1.32.1.vsix
92K	./ms-vscode.node-debug-1.35.2.vsix
616K	./ms-vscode.node-debug2-1.31.6.vsix
664K	./ms-vscode.node-debug2-1.33.0.vsix
3.3M	./netcoredbg_theia_plugin.theia
45M	./omnisharp_theia_plugin.theia
4.2M	./openshift-connector-0.0.17-127.vsix
4.5M	./openshift-connector-0.0.19-183.vsix
4.6M	./openshift-connector-0.0.21-228.vsix
1.1M	./php-debug.vsix
2.3M	./project_initializer-0.0.9-92.vsix
2.2M	./redhat.fabric8-analytics-0.0.12.vsix
5.6M	./redhat.vscode-xml-0.3.0.vsix
5.6M	./redhat.vscode-xml-0.5.1.vsix
5.8M	./redhat.vscode-yaml-0.3.0.vsix
5.8M	./redhat.vscode-yaml-0.4.0.vsix
35M	./sonarlint-vscode-1.6.0.vsix
19M	./vscode-apache-camel-0.0.16-37.vsix
3.7M	./vscode-java-debug-0.16.0.vsix
2.5M	./vscode-java-debug-0.19.0.vsix
17M	./vscode-kubernetes-tools-0.1.17.vsix
13M	./vscode-kubernetes-tools-1.0.0.vsix
472K	./vscode-pull-request-github-0.8.0.vsix
24M	./vscode-wsdl2rest-0.0.9-16.vsix
5.9M	./vscode-xml-0.7.0-3205.vsix
470M	total
```

nickboldt commented:
And:

```
  extensions:
  - relative:extension/resources/github.com/microsoft/vscode-java-debug/releases/download/0.19.0/vscode-java-debug-0.19.0.vsix
  - relative:extension/resources/download.jboss.org/jbosstools/static/jdt.ls/stable/java-0.46.0-1549.vsix
```

vs.

```
  extensions:
  - relative:extension/resources/vscode-java-debug-0.19.0.vsix
  - relative:extension/resources/java-0.46.0-1549.vsix
```

amisevsk commented:
@nickboldt Great catch, that would have been very hard to debug later.

We discussed this on slack but sharing it here for others to see:
- I've updated the PR to store extension artifacts in folders derived from their URL
- I also added a check to fail if we end up overwriting an already downloaded extension
- Finally, I figured out the source of the enormous image size (`chmod`ding 500Mi of files will explode image size, turns out). Current image size is `945MB`, which is `366MB` from the base httpd image, plus `579MB` in cached images.

<details><summary>Tree of files in `resources</summary>
<p>

```
$ tree /var/www/html/v3/resources
/var/www/html/v3/resources
|-- download.jboss.org
|   `-- jbosstools
|       |-- adapters
|       |   `-- stable
|       |       `-- vscode-project-initializer
|       |           `-- project_initializer-0.0.9-92.vsix
|       |-- static
|       |   `-- jdt.ls
|       |       `-- stable
|       |           |-- java-0.38.0-1373.vsix
|       |           |-- java-0.43.0-1473.vsix
|       |           |-- java-0.45.0-1523.vsix
|       |           `-- java-0.46.0-1549.vsix
|       `-- vscode
|           `-- stable
|               |-- vscode-apache-camel
|               |   `-- vscode-apache-camel-0.0.16-37.vsix
|               `-- vscode-wsdl2rest
|                   `-- vscode-wsdl2rest-0.0.9-16.vsix
|-- github.com
|   |-- Azure
|   |   `-- vscode-kubernetes-tools
|   |       `-- releases
|   |           `-- download
|   |               |-- 0.1.17
|   |               |   `-- vscode-kubernetes-tools-0.1.17.vsix
|   |               `-- 1.0.0
|   |                   `-- vscode-kubernetes-tools-1.0.0.vsix
|   |-- Microsoft
|   |   |-- vscode-go
|   |   |   `-- releases
|   |   |       `-- download
|   |   |           `-- 0.9.2
|   |   |               `-- Go-0.9.2.vsix
|   |   `-- vscode-python
|   |       `-- releases
|   |           `-- download
|   |               |-- 2019.3.6558
|   |               |   `-- ms-python-release.vsix
|   |               `-- 2019.5.18875
|   |                   `-- ms-python-release.vsix
|   |-- SonarSource
|   |   `-- sonarlint-vscode
|   |       `-- releases
|   |           `-- download
|   |               `-- 1.6.0
|   |                   `-- sonarlint-vscode-1.6.0.vsix
|   |-- camel-tooling
|   |   `-- camel-lsp-client-vscode
|   |       `-- releases
|   |           `-- download
|   |               `-- 0.0.14
|   |                   `-- camel-tooling.vscode-apache-camel-0.0.14.vsix
|   |-- che-incubator
|   |   |-- che-cpptools
|   |   |   `-- releases
|   |   |       `-- download
|   |   |           `-- v1.0
|   |   |               |-- cdt-gdb-vscode-0.0.90.vsix
|   |   |               |-- cdt-vscode-0.0.7.vsix
|   |   |               `-- llvm-vs-code-extensions.vscode-clangd-0.0.16.vsix
|   |   |-- ms-code.typescript
|   |   |   `-- releases
|   |   |       `-- download
|   |   |           |-- v1.30.2
|   |   |           |   `-- che-typescript-language.vsix
|   |   |           `-- v1.35.1
|   |   |               `-- che-typescript-language-1.35.1.vsix
|   |   `-- vscode-intelephense
|   |       `-- releases
|   |           `-- download
|   |               `-- v1.0.13
|   |                   `-- bmewburn.vscode-intelephense-client-1.0.13.vsix
|   |-- fabric8-analytics
|   |   `-- fabric8-analytics-vscode-extension
|   |       `-- releases
|   |           `-- download
|   |               `-- 0.0.12
|   |                   `-- redhat.fabric8-analytics-0.0.12.vsix
|   |-- felixfbecker
|   |   `-- vscode-php-debug
|   |       `-- releases
|   |           `-- download
|   |               `-- v1.13.0
|   |                   `-- php-debug.vsix
|   |-- microsoft
|   |   |-- vscode-go
|   |   |   `-- releases
|   |   |       `-- download
|   |   |           `-- 0.11.0
|   |   |               `-- Go-0.11.0.vsix
|   |   |-- vscode-java-debug
|   |   |   `-- releases
|   |   |       `-- download
|   |   |           |-- 0.16.0
|   |   |           |   `-- vscode-java-debug-0.16.0.vsix
|   |   |           `-- 0.19.0
|   |   |               `-- vscode-java-debug-0.19.0.vsix
|   |   `-- vscode-pull-request-github
|   |       `-- releases
|   |           `-- download
|   |               `-- v0.8.0
|   |                   `-- vscode-pull-request-github-0.8.0.vsix
|   `-- redhat-developer
|       |-- netcoredbg-theia-plugin
|       |   `-- releases
|       |       `-- download
|       |           `-- v0.0.1
|       |               `-- netcoredbg_theia_plugin.theia
|       |-- omnisharp-theia-plugin
|       |   `-- releases
|       |       `-- download
|       |           |-- v0.0.1
|       |           |   `-- omnisharp_theia_plugin.theia
|       |           |-- v0.0.2
|       |           |   `-- omnisharp_theia_plugin.theia
|       |           `-- v0.0.4
|       |               `-- omnisharp_theia_plugin.theia
|       |-- vscode-openshift-tools
|       |   `-- releases
|       |       `-- download
|       |           |-- v0.0.17
|       |           |   `-- openshift-connector-0.0.17-127.vsix
|       |           |-- v0.0.19
|       |           |   `-- openshift-connector-0.0.19-183.vsix
|       |           `-- v0.0.21
|       |               `-- openshift-connector-0.0.21-228.vsix
|       |-- vscode-xml
|       |   `-- releases
|       |       `-- download
|       |           |-- 0.3.0
|       |           |   `-- redhat.vscode-xml-0.3.0.vsix
|       |           |-- 0.5.1
|       |           |   `-- redhat.vscode-xml-0.5.1.vsix
|       |           `-- 0.7.0
|       |               `-- vscode-xml-0.7.0-3205.vsix
|       `-- vscode-yaml
|           `-- releases
|               `-- download
|                   |-- 0.3.0
|                   |   `-- redhat.vscode-yaml-0.3.0.vsix
|                   `-- 0.4.0
|                       `-- redhat.vscode-yaml-0.4.0.vsix
`-- marketplace.visualstudio.com
    `-- _apis
        `-- public
            `-- gallery
                `-- publishers
                    |-- joaompinto
                    |   `-- vsextensions
                    |       `-- asciidoctor-vscode
                    |           `-- 2.7.6
                    |               `-- joaompinto.asciidoctor-vscode-2.7.6.vsix
                    |-- ms-python
                    |   `-- vsextensions
                    |       `-- python
                    |           `-- 2019.2.5433
                    |               `-- ms-python.python-2019.2.5433.vsix
                    `-- ms-vscode
                        `-- vsextensions
                            |-- node-debug
                            |   |-- 1.32.1
                            |   |   `-- ms-vscode.node-debug-1.32.1.vsix
                            |   `-- 1.35.2
                            |       `-- ms-vscode.node-debug-1.35.2.vsix
                            `-- node-debug2
                                |-- 1.31.6
                                |   `-- ms-vscode.node-debug2-1.31.6.vsix
                                `-- 1.33.0
                                    `-- ms-vscode.node-debug2-1.33.0.vsix

126 directories, 44 files
```

</p>
</details

amisevsk commented:
If you don't want to sit through the build process, the image built from the last commit is available at `amisevsk/che-plugin-registry:offline-mode`

benoitf commented:
maybe for offline mode we should only keep latest version only (as it's for productization I'm not sure we want to maintain every versions of vscode java, omnisharp, etc).

I see 4 versions of vscode-java, 3 of omnisharp, etc.

as a side note it seems we're still using marketplace as I'm seeing in folder:
```
sh-4.2$ ls marketplace.visualstudio.com/_apis/public/gallery/publishers/*
marketplace.visualstudio.com/_apis/public/gallery/publishers/joaompinto:
vsextensions

marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python:
vsextensions

marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-vscode:
vsextensions

```

amisevsk commented:
> maybe for offline mode we should only keep latest version only

We can't do that since old already-created workspaces could depend on the old versions.

We are still using the marketplace, since a few extensions are still not publishing .vsix files anywhere else.

nickboldt commented:
For the product fork of this, I will be pulling only latest and as to building vsix files that are only on marketplace, see https://github.com/eclipse/che/issues/14573 - @mkuznyetsov is working on setting up a job that can build the ms-node-debug stuff. Once we're building it, we can extend that to build all the latest versions and use the built-by-redhat versions in the product.

For upstream Che 7.3, we can publish our own built-by-redhat versions onto download.jboss.org (as we already do for the stuff we own) so that we can refer to them online but from a safe URL location.

nickboldt commented on dockerfile-test:
```diff
@@ -0,0 +1,4 @@
+FROM amisevsk/che-plugin-registry:offline-mode
```
do we need this file committed? Seems like it's less a test script and more a thing-you-used-for-manual-verification :)

amisevsk replied:
It suuure is :)

nickboldt commented on Dockerfile.rhel:
```diff
@@ -31,9 +34,19 @@ EXPOSE 80
+FROM registry AS offline-registry
+COPY --from=offline-builder /build/v3 /var/www/html/v3
+USER 0
+RUN chmod -R g+rwX /var/www/html/v3/
```
can you add a `WORKDIR /var/www/html/` to the end of the Dockerfile* files so that if you connect to the image you're already in the folder where all the magic happens? 

(rather than being in `/opt/app-root/src` where nothing magical happens)

amisevsk commented on Dockerfile:
```diff
@@ -6,20 +6,31 @@
+
+# Offline build: cache .theia and .vsix files in registry itself and update metas
+FROM builder AS offline-builder
+RUN ./cache_artifacts.sh v3
```
It's confusing reading the dockerfile to have `chmod` in the builder and not in the offline-builder. I prefer it as a layer here.

amisevsk commented:
Squashed the messy commit history
